### PR TITLE
[ASA-257][RV] add check-answers page

### PR DESCRIPTION
**Pull request by rvijayan**

### CODE REVIEW

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.controllers.auth.SubscriptionAuthAction
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.views.html._
+
+import scala.concurrent.Future
+
+@Singleton
+class BusinessIdentificationController @Inject()(
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: we do not need this controller.

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  val messagesApi: MessagesApi,
+  subscriptionAuthAction: SubscriptionAuthAction)(implicit val configuration: Configuration)
+    extends FrontendController with I18nSupport {
+
+  def showCheckAnswers: Action[AnyContent] = subscriptionAuthAction.async { implicit request =>
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: the "implicit request" should be "implicit SubscriptionApplicationRequest"

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/auth/HmrcAsAgentAuthAction.scala [hunk 3]
```diff
@@ -46,7 +49,7 @@ class HmrcAsAgentAuthAction @Inject()(val env: Environment, val authConnector: A
             .map(Arn(_))
             .getOrElse(throw new InsufficientEnrolments("could not find the HMRC-AS-AGENT enrolment to continue"))
 
-          block(HmrcAsAgentRequest(arn, request))
+          block(ArnRequest(arn, request))
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/auth/HmrcAsAgentAuthAction.scala]: ArnRequest can mean many things, HmrcAsAgentRequest I believe was clear, that it is a request that contains that particular enrolment...

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/models/BusinessAddress.scala [hunk 4]
```diff
@@ -22,6 +22,7 @@ case class BusinessAddress(
   addressline1: String,
   addressline2: Option[String],
   addressline3: Option[String],
+  country: String,
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/models/BusinessAddress.scala]: this data is not provided?

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/views/check_answers.scala.html [hunk 5]
```diff
@@ -15,10 +15,67 @@
  *@
 
 @import play.api.Configuration
+@import uk.gov.hmrc.agentoverseassubscriptionfrontend.models.SubscriptionApplication
 
-@()(implicit messages: Messages, request : Request[_], configuration: Configuration)
+@(application: SubscriptionApplication)(implicit messages: Messages, request : Request[_], configuration: Configuration)
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/views/check_answers.scala.html]: Somme will review all the pages when were done with all this

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import play.api.libs.ws.WSClient
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.stubs.AgentOverseasApplicationStubs
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.support.BaseISpec
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.support.SampleUsers.authenticatedCleanCred
+
+class BusinessIdentificationControllerISpec extends BaseISpec with AgentOverseasApplicationStubs {
+
+  implicit val ws: WSClient = app.injector.instanceOf(classOf[WSClient])
+
+  private def appBaseUrl = s"/agent-services/apply-from-outside-uk/create-account"
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala]: we will convert this appbaseUrl to use routes.<controllerName>.<function>().url it is identical and does not require maintance

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+  "GET /check-answers" should {
+    "display the check-answers page" in {
+      givenUserIsAuthenticated(authenticatedCleanCred)
+      givenSubscriptionApplicationResponse("12345-credId")
+
+      val result = wsUrl(s"$appBaseUrl/check-answers").get().futureValue
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala]: the above comment will go here and the appBaseUrl will be removed

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.controllers.auth.SubscriptionAuthAction
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.models.{ApplicationStatus, BusinessAddress, SubscriptionApplication, SubscriptionApplicationRequest}
+import uk.gov.hmrc.agentoverseassubscriptionfrontend.support.BaseUnitSpec
+
+import scala.concurrent.Future
+
+class BusinessIdentificationControllerSpec extends BaseUnitSpec {
+
+  private val application = SubscriptionApplication(
+    "SomeId",
```
Comment (arkowalski) [inline test/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerSpec.scala]: if you know what the ID will look like update else ignore this comment

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+trait BaseUnitSpec extends UnitSpec with WithFakeApplication with MockitoSugar {
+
+  def injector: Injector = fakeApplication.injector
+  implicit val mockConfig = injector.instanceOf[Configuration]
+
+  protected implicit val materializer = fakeApplication.materializer
+  implicit val duration: Timeout = 20 seconds
+
+  val messagesApi = injector.instanceOf[MessagesApi]
```
Comment (arkowalski) [inline test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala]: should this be mocked?

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  private implicit val messages: Messages = messagesApi.preferred(Seq.empty[Lang])
+
+  protected def htmlEscapedMessage(key: String, args: Any*): String = escape(Messages(key, args: _*)).toString
+
+  implicit def hc(implicit request: FakeRequest[_]): HeaderCarrier =
+    HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+  protected def checkMessageIsDefined(messageKey: String) =
```
Comment (arkowalski) [inline test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala]: can we make this private, so we can just use containSubstrings

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+        MatchResult(
+          strsMissing.isEmpty,
+          s"Expected substrings are missing in the response: ${strsMissing.mkString("\"", "\", \"", "\"")}",
+          s"Expected substrings are present in the response : ${strsPresent.mkString("\"", "\", \"", "\"")}"
+        )
+      }
+    }
+
+  protected def containMessages(expectedMessageKeys: String*): Matcher[Result] =
```
Comment (arkowalski) [inline test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala]: why is this here? Looks like same on line 61

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala (hunk 1 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: This controller is for all the actions required for showing and chaing business details. Its similar to what we have done in subscription frontend service.

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala (hunk 2 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: subscriptionAuthAction will return a SubscriptionApplicationRequest for sure. So I dont think we need to explicitly state so..

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/auth/HmrcAsAgentAuthAction.scala (hunk 3 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/auth/HmrcAsAgentAuthAction.scala]: Well, I thought HmrcAsAgent is an enrolment and the Request does not contain any info related to that enrolment other than ARN. I just followed the naming pattern we did for other requests.

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/models/BusinessAddress.scala (hunk 4 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/models/BusinessAddress.scala]: This data should come from the application. its required to show country in the check-answers page.

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala (hunk 6 above)
Comment (rvijayan) [reply it/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala]: It will allow us to minimise errors if you explicitly mention urls in tests.

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala (hunk 10 above)
Comment (rvijayan) [reply test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala]: we could use this method for some tests in future if is public.

File: test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala (hunk 11 above)
Comment (rvijayan) [reply test/uk/gov/hmrc/agentoverseassubscriptionfrontend/support/BaseUnitSpec.scala]: Its different. The method on line 61 is for sub strings.

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala (hunk 2 above)
Comment (arkowalski) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: its much more clearer to see that you have access to more data

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala (hunk 1 above)
Comment (arkowalski) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/controllers/BusinessIdentificationController.scala]: we have bunch of things happening in BusinessidentificationController in agent-sub-FE and were not doing any crazy things, one controller should be enough as it is all related. only think I can think of is the clean creds so a signout controller for that maybe
